File: apps/web/src/shared/monitorApi.ts
import { apiClient } from './apiClient';

export type MonitorType = 'HTTPS' | 'HTTP' | 'SSL' | 'TCP' | 'DNS';
export type MonitorStatus = 'UP' | 'DOWN' | 'UNKNOWN';

export type Monitor = {
  id: number;
  name: string;
  type: MonitorType;
  target: string;
  expectedStatusCode: number;
  frequencySeconds: number;
  timeoutSeconds: number;
  currentStatus: MonitorStatus;
  lastResponseTime?: number | null;
  lastCheckedAt?: string | null;
  nextCheckAt?: string;
  isActive: boolean;
  usesSectionSchedule?: boolean;
  locations: string[];
  alertEmail: boolean;
  alertPush: boolean;
  alertThreshold: number;
  tcpPort?: number | null;
  keyword?: string | null;
  sslWarningDays?: number | null;
  dnsRecordType?: string | null;
  dnsExpectedValue?: string | null;
  sections?: {
    section: {
      id: number;
      name: string;
      expectedStatusCode: number;
      frequencySeconds: number;
      timeoutSeconds: number;
      locations: string[];
      isActive: boolean;
    };
  }[];
};

export type MonitorCheck = {
  id: number;
  monitorId: number;
  status: MonitorStatus;
  responseTimeMs: number | null;
  statusCode: number | null;
  errorMessage: string | null;
  location?: string | null;
  checkedAt: string;
};

export type PaginatedMonitors = {
  items: Monitor[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
};

export type MonitorSortOption = 'status' | 'name' | 'latest-check' | 'created-at';
export type MonitorViewStatusFilter = 'ALL' | 'UP' | 'DOWN' | 'PAUSED' | 'UNKNOWN';

export type MonitorListQuery = {
  page?: number;
  limit?: number;
  search?: string;
  sort?: MonitorSortOption;
  status?: MonitorViewStatusFilter;
  type?: 'ALL' | MonitorType;
  location?: string;
};

export type CreateMonitorInput = {
  name: string;
  type: MonitorType;
  target: string;
  expectedStatusCode: number;
  frequencySeconds: number;
  timeoutSeconds: number;
  locations: string[];
  alertEmail: boolean;
  alertPush: boolean;
  alertThreshold: number;
  tcpPort?: number | null;
  keyword?: string | null;
  sslWarningDays?: number | null;
  dnsRecordType?: string | null;
  dnsExpectedValue?: string | null;
};

export type UpdateMonitorInput = CreateMonitorInput;

const MONITOR_LIST_FETCH_LIMIT = 100;

function buildMonitorListPath(query: MonitorListQuery = {}) {
  const params = new URLSearchParams();

  for (const [key, rawValue] of Object.entries(query)) {
    if (rawValue === undefined || rawValue === '' || rawValue === 'ALL') {
      continue;
    }

    params.set(key, String(rawValue));
  }

  const queryString = params.toString();
  return `/monitors${queryString ? `?${queryString}` : ''}`;
}

export const getPaginatedMonitors = async (
  query: MonitorListQuery = {},
  signal?: AbortSignal,
) => {
  return apiClient<PaginatedMonitors>(buildMonitorListPath(query), { signal });
};

export const getMonitors = async () => {
  const firstPage = await getPaginatedMonitors({
    limit: MONITOR_LIST_FETCH_LIMIT,
    page: 1,
  });

  if (firstPage.totalPages <= 1) {
    return firstPage.items;
  }

  const remainingPages = await Promise.all(
    Array.from({ length: firstPage.totalPages - 1 }, (_, index) =>
      getPaginatedMonitors({
        limit: MONITOR_LIST_FETCH_LIMIT,
        page: index + 2,
      }),
    ),
  );

  return [
    ...firstPage.items,
    ...remainingPages.flatMap((page) => page.items),
  ];
};

export const createMonitor = async (data: CreateMonitorInput) => {
  return apiClient<Monitor>('/monitors', {
    method: 'POST',
    body: JSON.stringify(data),
  });
};

export const updateMonitor = async (id: number, data: UpdateMonitorInput) => {
  return apiClient<Monitor>(`/monitors/${encodeURIComponent(String(id))}`, {
    method: 'PATCH',
    body: JSON.stringify(data),
  });
};

export const useSectionSchedule = async (id: number) => {
  return apiClient<Monitor>(`/monitors/${encodeURIComponent(String(id))}/use-section-schedule`, {
    method: 'PATCH',
  });
};

export const deleteMonitor = async (id: number) => {
  return apiClient<Monitor>(`/monitors/${encodeURIComponent(String(id))}`, {
    method: 'DELETE',
  });
};

export const runMonitorCheck = async (id: number) => {
  return apiClient(`/monitors/${encodeURIComponent(String(id))}/run-check`, {
    method: 'POST',
  });
};

export const toggleMonitorActive = async (id: number) => {
  return apiClient<Monitor>(`/monitors/${encodeURIComponent(String(id))}/toggle-active`, {
    method: 'PATCH',
  });
};

export const getMonitor = async (id: number) => {
  return apiClient<Monitor>(`/monitors/${encodeURIComponent(String(id))}`);
};

export const getMonitorChecks = async (
  id: number,
  order: 'asc' | 'desc' = 'desc',
) => {
  return apiClient<MonitorCheck[]>(`/monitors/${encodeURIComponent(String(id))}/checks?order=${encodeURIComponent(order)}`);
};

File: apps/mobile/lib/features/monitors/data/monitor_api.dart
import '../../../core/network/api_client.dart';

class MonitorApi {
  MonitorApi({ApiClient? client}) : _client = client ?? ApiClient();

  final ApiClient _client;

  Future<List<Monitor>> getMonitors() async {
    final data = await _client.get('/monitors');
    return _asList(data).map(Monitor.fromJson).toList();
  }

  Future<Monitor> getMonitorById(int id) async {
    final data = await _client.get('/monitors/$id');
    return Monitor.fromJson(_asMap(data));
  }

  Future<Monitor> createMonitor(Map<String, dynamic> data) async {
    final response = await _client.post('/monitors', body: data);
    return Monitor.fromJson(_asMap(response));
  }

  Future<Monitor> updateMonitor(int id, Map<String, dynamic> data) async {
    final response = await _client.patch('/monitors/$id', body: data);
    return Monitor.fromJson(_asMap(response));
  }

  Future<void> deleteMonitor(int id) async {
    await _client.delete('/monitors/$id');
  }

  Future<dynamic> runCheck(int id) async {
    try {
      return await _client.post('/monitors/$id/check');
    } on ApiException catch (error) {
      if (error.statusCode == 404) {
        return _client.post('/monitors/$id/run-check');
      }
      rethrow;
    }
  }
}

class Monitor {
  const Monitor({
    required this.id,
    required this.name,
    required this.type,
    required this.target,
    required this.currentStatus,
    required this.isActive,
    this.expectedStatusCode,
    this.frequencySeconds,
    this.timeoutSeconds,
    this.lastResponseTime,
    this.lastCheckedAt,
    this.checkResults = const [],
  });

  final int id;
  final String name;
  final String type;
  final String target;
  final String currentStatus;
  final bool isActive;
  final int? expectedStatusCode;
  final int? frequencySeconds;
  final int? timeoutSeconds;
  final int? lastResponseTime;
  final DateTime? lastCheckedAt;
  final List<CheckResult> checkResults;

  factory Monitor.fromJson(Map<String, dynamic> json) {
    return Monitor(
      id: _asInt(json['id']),
      name: (json['name'] ?? '').toString(),
      type: (json['type'] ?? '').toString(),
      target: (json['target'] ?? '').toString(),
      currentStatus: (json['currentStatus'] ?? 'UNKNOWN').toString(),
      isActive: json['isActive'] != false,
      expectedStatusCode: _asNullableInt(json['expectedStatusCode']),
      frequencySeconds: _asNullableInt(json['frequencySeconds']),
      timeoutSeconds: _asNullableInt(json['timeoutSeconds']),
      lastResponseTime: _asNullableInt(json['lastResponseTime']),
      lastCheckedAt: _asDateTime(json['lastCheckedAt']),
      checkResults: _asList(
        json['checkResults'],
      ).map(CheckResult.fromJson).toList(),
    );
  }
}

class CheckResult {
  const CheckResult({
    required this.id,
    required this.status,
    this.responseTimeMs,
    this.statusCode,
    this.errorMessage,
    this.location,
    this.checkedAt,
  });

  final int id;
  final String status;
  final int? responseTimeMs;
  final int? statusCode;
  final String? errorMessage;
  final String? location;
  final DateTime? checkedAt;

  factory CheckResult.fromJson(Map<String, dynamic> json) {
    return CheckResult(
      id: _asInt(json['id']),
      status: (json['status'] ?? '').toString(),
      responseTimeMs: _asNullableInt(json['responseTimeMs']),
      statusCode: _asNullableInt(json['statusCode']),
      errorMessage: json['errorMessage']?.toString(),
      location: json['location']?.toString(),
      checkedAt: _asDateTime(json['checkedAt']),
    );
  }
}

List<Map<String, dynamic>> _asList(dynamic value) {
  if (value is List) {
    return value.map(_asMap).toList();
  }
  return const [];
}

Map<String, dynamic> _asMap(dynamic value) {
  if (value is Map<String, dynamic>) return value;
  if (value is Map) return Map<String, dynamic>.from(value);
  return <String, dynamic>{};
}

int _asInt(dynamic value) {
  return _asNullableInt(value) ?? 0;
}

int? _asNullableInt(dynamic value) {
  if (value is int) return value;
  if (value is num) return value.toInt();
  return int.tryParse(value?.toString() ?? '');
}

DateTime? _asDateTime(dynamic value) {
  if (value == null) return null;
  return DateTime.tryParse(value.toString());
}
